accepted/wasmtime-compile-time-builtins.md

```diff
@@ -0,0 +1,399 @@
+instructions.
+
+First up are load and store operations for the native address space. The
+`pointer` type is either an `i32` or `i64` depending on the target's pointer
+width. The `native` mnemonic means that the memory operation is operating on the
+native memory address space, not a particular Wasm memory, and uses native
+endianness. The `_s` and `_u` mnemonics have the same meaning as in core Wasm:
+they specify whether the value is signed- or unsigned extended. The `8` and `16`
+suffixes also have the same meaning as in core Wasm: only store the bottom N
+bits of the value to memory.
+
+* `i8.native_load_s: [pointer] -> [i32]`
+* `i8.native_load_u: [pointer] -> [i32]`
+* `i16.native_load_s: [pointer] -> [i32]`
+* `i16.native_load_u: [pointer] -> [i32]`
+* `i32.native_load: [pointer] -> [i32]`
+* `i64.native_load: [pointer] -> [i64]`
+* `i32.native_store8: [pointer i32] -> []`
+* `i32.native_store16: [pointer i32] -> []`
+* `i32.native_store: [pointer i32] -> []`
+* `i64.native_store: [pointer i64] -> []`
+
+Next, we have an intrinsic for getting the address of the host data in the
+resource table for a particular `i32` resource handle. It will raise a trap on
+out-of-bounds resource table accesses.
+
+* `resource.address: [i32] -> [pointer]`
+
+Note that we do *not* define intrinsics for directly accessing or addressing the
+`vmctx`, any linear memories, or any other internal state of the Wasm
+instance. While Wasmtime needs those abilities to implement operations like
+`global.get`, compile-time builtins do not get to see inside Wasmtime's
+implementation details. They are only given helpers for accessing things that
+the embedder themselves defined, such as the elements inside a resource
+table.[^blah]
+
+[^blah]: Of course, although they should only ever access embedder-defined data
+    as part of their safety contract, compile-time builtins ultimately have the
+    capability to access absolutely anything -- including Wasmtime internals --
+    since they have native address space loads and stores. That doesn't mean we
+    need to add more footguns than we fundamentally must.
+
+While we needn't implement all of these intrinsics from the very start, we
+should in the fullness of time implement all of them.
+
+# Open questions
+[open-questions]: #open-questions
+
+* Is there some other approach to defining compile-time builtins for inlining
+  that I am failing to identify and which is better than anything proposed here?
+
+* How do we specify which resource table we want to access in the
+  `resource.address` intrinsic? Do we need separate mechanisms for defining core
+  and component compile-time builtins, where component builtins are self-hosted
+  components and core builtins are self-hosted core modules?
+
+* Should we even support defining compile-time builtins with Winch at all? Or
+  should it always make the function calls at runtime, since if you wanted to
+  prioritize runtime speed you'd be using Cranelift anyways?
```

Comment:
I believe the Winch topic was (briefly?) discussed during the [Wasmtime 06-05 meeting](https://github.com/bytecodealliance/meetings/pull/618). However, I wanted to mention that I agree with the no-inlining direction for Winch; even though I believe that some sort of inlining is feasible, especially for the types of functions outlined in this RFC, I'm also very confident that doing so, has the potential to against Winch's simplicity and compilation performance design principles.

Reply:
Makes total sense, thanks for weighing in.